pkg/bloomgateway/client.go

```diff
@@ -0,0 +1,287 @@
+}
+
+type GatewayClient struct {
+	cfg    ClientConfig
+	limits Limits
+	logger log.Logger
+	pool   *ringclient.Pool
+	ring   ring.ReadRing
+}
+
+func NewGatewayClient(cfg ClientConfig, limits Limits, registerer prometheus.Registerer, logger log.Logger) (*GatewayClient, error) {
+	latency := promauto.With(registerer).NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "loki",
+		Subsystem: "bloom_gateway",
+		Name:      "request_duration_seconds",
+		Help:      "Time (in seconds) spent serving requests when using the bloom gateway",
+		Buckets:   instrument.DefBuckets,
+	}, []string{"operation", "status_code"})
+
+	dialOpts, err := cfg.GRPCClientConfig.DialOption(grpcclient.Instrument(latency))
+	if err != nil {
+		return nil, err
+	}
+
+	poolFactory := func(addr string) (ringclient.PoolClient, error) {
+		pool, err := NewBloomGatewayGRPCPool(addr, dialOpts)
+		if err != nil {
+			return nil, errors.Wrap(err, "new bloom gateway grpc pool")
+		}
+		return pool, nil
+	}
+
+	c := &GatewayClient{
+		cfg:    cfg,
+		logger: logger,
+		limits: limits,
+		pool:   clientpool.NewPool(cfg.PoolConfig, cfg.Ring, poolFactory, logger),
+	}
+
+	return c, nil
+}
+
+func shuffleAddrs(addrs []string) []string {
+	rand.Shuffle(len(addrs), func(i, j int) {
+		addrs[i], addrs[j] = addrs[j], addrs[i]
+	})
+	return addrs
+}
+
+// FilterChunkRefs implements Client
+func (c *GatewayClient) FilterChunks(ctx context.Context, tenant string, from, through model.Time, fingerprints []uint64, chunkRefs [][]*logproto.ChunkRef, filters ...*logproto.LineFilterExpression) ([]uint64, [][]*logproto.ChunkRef, error) {
+	// Get the addresses of corresponding bloom gateways for each series.
+	_, addrs, err := c.serverAddrsForFingerprints(tenant, fingerprints)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Group chunk refs by addresses of one or more bloom gateways.
+	// All chunk refs of series that belong to one and the same bloom gateway are set in one batch.
+	streamsByAddr := c.groupStreamsByAddr(fingerprints, chunkRefs, addrs)
+
+	// TODO(chaudum): We might over-allocate for the filtered responses here?
+	filteredChunkRefs := make([][]*logproto.ChunkRef, 0, len(fingerprints))
+	filteredFingerprints := make([]uint64, 0, len(fingerprints))
+
+	for _, item := range streamsByAddr {
+		// randomize order of addresses so we don't hotspot the first server in the list
+		addrs := shuffleAddrs(item.addrs)
+		err := c.doForAddrs(addrs, func(client logproto.BloomGatewayClient) error {
+			req := &logproto.FilterChunkRefRequest{
+				From:    from,
+				Through: through,
+				Refs:    item.refs,
+				Filters: filters,
+			}
+			resp, err := client.FilterChunkRefs(ctx, req)
+			if err != nil {
+				return err
+			}
+			for _, res := range resp.Chunks {
+				chunkRefs := make([]*logproto.ChunkRef, 0, len(res.ChunkIDs))
+				for _, chunkID := range res.ChunkIDs {
+					chk, err := chunk.ParseExternalKey(tenant, chunkID)
```

Comment:
It probably makes more sense to send the chunkref structs directly in protos rather than parse them on either end. This may be more efficient and removes these error conditions.